src/ops/fuelup_self.rs

```diff
@@ -1,17 +1,70 @@
-use anyhow::Result;
+use anyhow::{bail, Result};
+use std::{fs, path::Path};
+use tempfile::tempdir_in;
+use tracing::{error, info};
 
 use crate::{
     component,
     download::{download_file_and_unpack, unpack_bins, DownloadCfg},
-    path::fuelup_bin_dir,
+    path::{fuelup_bin, fuelup_bin_dir},
 };
 
+pub fn attempt_install_self(download_cfg: DownloadCfg, dst: &Path) -> Result<()> {
+    download_file_and_unpack(&download_cfg, dst)?;
+    unpack_bins(dst, dst)?;
+
+    Ok(())
+}
+
 pub fn self_update() -> Result<()> {
     let download_cfg = DownloadCfg::new(component::FUELUP, None)?;
+
+    let fuelup_bin = fuelup_bin();
     let fuelup_bin_dir = fuelup_bin_dir();
+    let fuelup_backup = fuelup_bin_dir.join("fuelup-backup");
+
+    let fuelup_new_dir = tempdir_in(&fuelup_bin_dir)?;
+    let fuelup_new_dir_path = fuelup_new_dir.path();
+
+    if let Err(e) = attempt_install_self(download_cfg, fuelup_new_dir_path) {
+        // Skip all other steps if downloading fails here.
+        // We do not need to handle failure, since this downloads to a tempdir.
+        bail!("Failed to install fuelup: {}", e);
+    };
+
+    if fuelup_bin.exists() {
+        // Make a backup of fuelup, fuelup-backup.
+        if let Err(e) = fs::rename(&fuelup_bin, &fuelup_backup) {
+            bail!(
+                "Failed moving {} to {}: {}",
+                &fuelup_bin.display(),
+                &fuelup_backup.display(),
+                e
+            );
+        }
+    };
+
+    info!(
+        "Moving {} to {}",
+        fuelup_new_dir_path.join("fuelup").display(),
+        &fuelup_bin.display()
+    );
+    if let Err(e) = fs::rename(fuelup_new_dir_path.join("fuelup"), &fuelup_bin) {
+        error!("Failed to replace the old fuelup: {}", e);
+        error!("Restoring old fuelup");
+
+        // If we have failed to replace the old fuelup for whatever reason, we want the backup.
+        // Although unlikely, should this last step fail, we will recommend to re-install fuelup using fuelup-init.
+        if let Err(e) = fs::rename(&fuelup_backup, &fuelup_bin) {
+            error!("Could not restore backup fuelup: {}", e);
+            error!("You should re-install fuelup using fuelup-init:");
+            error!("`curl --proto '=https' --tlsv1.2 -sSf https://fuellabs.github.io/fuelup/fuelup-init.sh | sh`");
+            std::process::exit(1);
+        }
```

Comment:
A couple things here:

1. We should avoid adding a `std::process::exit` in the middle of an unsuspecting function like `self_update`. `std::process::exit` terminates the program and does not return, which might be confusing for a future maintainer if we want to refactor error handling or something in the future. I think it's better practice to restrict the use of `exit` to the very top-level, e.g. `main` or `run` if `main` is unsuitable for whatever reason.

2. This still returns a success exit code in the case that the rename from `fuelup_new_dir_path.join("fuelup")` to `fuelup_bin` fails. I'm not sure this is what we want?

3. Rather than using `error!` (here and above) and `exit`, perhaps we can format this into an anyhow error that can be returned to keep error handling more consistent?

Reply:
Hmm, at first the thought was that the `error!`s are there to provide additional info on what `fuelup` is trying to do if the `fs::rename()`s fails. But you're right in that we can just use `bail!` instead with the `error!`s to exit out of the function - updated it to do this instead, with only 1 usage of `error!`